Build work

wadells opened the pull request:
This PR includes improvements to the build:
 - python packaging
 - with a tar of the wellknown container baked into the python build
 - improved cleanup of dev artifacts (i.e. `make prune` to clear out leftover images from failed test runs)

See individual commit messages for more info.

Testing Done:
```
# clean and full build

(venv) vagrant@ubuntu-bionic:/vagrant$ make clean
rm -f src/grav/wellknown.tar wellknown-container/.image_id wellknown-container/.container_id dist/gravitational_audition-0.1.0-py3-none-any.whl
(venv) vagrant@ubuntu-bionic:/vagrant$ make build
docker build  --iidfile wellknown-container/.image_id wellknown-container
Sending build context to Docker daemon  3.072kB
Step 1/3 : from alpine:3.10
 ---> 961769676411
Step 2/3 : copy loop.sh /tmp/loop.sh
 ---> Using cache
 ---> 82e4febc0d79
Step 3/3 : cmd ["sh", "/tmp/loop.sh"]
 ---> Using cache
 ---> a34690973417
Successfully built a34690973417
docker save --output src/grav/wellknown.tar sha256:a34690973417bdec5f1056a398a7bb4223f64dc05820e2c407695e9937e76340
pip wheel --wheel-dir dist/ .
Processing /vagrant
Building wheels for collected packages: gravitational-audition
  Running setup.py bdist_wheel for gravitational-audition ... done
  Stored in directory: /vagrant/dist
Successfully built gravitational-audition


# partial build with nothing changed

(venv) vagrant@ubuntu-bionic:/vagrant$ make build
make: Nothing to be done for 'build'.


# partial build with the container changed

(venv) vagrant@ubuntu-bionic:/vagrant$ touch wellknown-container/loop.sh
(venv) vagrant@ubuntu-bionic:/vagrant$ make build
docker build  --iidfile wellknown-container/.image_id wellknown-container
Sending build context to Docker daemon  3.072kB
Step 1/3 : from alpine:3.10
 ---> 961769676411
Step 2/3 : copy loop.sh /tmp/loop.sh
 ---> Using cache
 ---> 82e4febc0d79
Step 3/3 : cmd ["sh", "/tmp/loop.sh"]
 ---> Using cache
 ---> a34690973417
Successfully built a34690973417
docker save --output src/grav/wellknown.tar sha256:a34690973417bdec5f1056a398a7bb4223f64dc05820e2c407695e9937e76340
pip wheel --wheel-dir dist/ .
Processing /vagrant
Building wheels for collected packages: gravitational-audition
  Running setup.py bdist_wheel for gravitational-audition ... done
  Stored in directory: /vagrant/dist
Successfully built gravitational-audition
(venv) vagrant@ubuntu-bionic:/vagrant$ 


# pruning docker images
(venv) vagrant@ubuntu-bionic:/vagrant$ docker image list
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
<none>              <none>              a34690973417        25 minutes ago      5.58MB
alpine              3.10                961769676411        8 weeks ago         5.58MB
(venv) vagrant@ubuntu-bionic:/vagrant$ make prune
rm -f src/grav/wellknown.tar wellknown-container/.image_id wellknown-container/.container_id dist/gravitational_audition-0.1.0-py3-none-any.whl
docker system prune -f
Deleted Images:
deleted: sha256:a34690973417bdec5f1056a398a7bb4223f64dc05820e2c407695e9937e76340
deleted: sha256:82e4febc0d792bd19806f87a4475f6dfcf010da9033984ea4cad5de5edceec78
deleted: sha256:858bde05b018398fb09bbb24fd59f7531511a54a9afdbcec8eabbbfe03f65e02

Total reclaimed space: 308B
(venv) vagrant@ubuntu-bionic:/vagrant$ docker image list
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
alpine              3.10                961769676411        8 weeks ago         5.58MB
```

wadells committed:
Add initial python packaging.

One of the requests in the challenge doc was reproduceable builds.
We're gonna use the standard python build tool chain.  Nothing fancy.

Added a couple .gitignore lines related to python build artifacts.

wadells committed:
Add a tar of the wellknown image to the python distribution.

In order to accomplish my goal of "test code doesn't need to hit the
network" I need a way for the test to get a copy of the container that
should be tested.  Thankfully docker can import from a tar!

Thus as part of the build we generate wellknown.tar, and add that to the
python packaging manifest.

Since this is an intermediate build artifact, it gets .gitignored.

wadells committed:
Fix package_dir in setup.py.

I mistook the syntax of this, and overspecified the grav package.

The generated artifact actually contains all the right files now.

wadells committed:
Add more python packaging artifacts to .gitignore.

wadells committed:
Add 'prune' target to Makefile.

During testing, I found that I was accumulating a bunch of stray
containers and images from aborted test attempts.  This make target
gives a quick shorthand to clean those up, as well as the tracking
cookies that make remembers container/image IDs in.

wadells committed:
Ignore .whl files too.

Yet another python packaging artifact.

wadells committed:
Add build target.

Now that we produce a python package, we assign that `make build`.

This chains off the docker image, since that is baked into the package.

r0mant commented on Makefile:
```diff
@@ -26,5 +46,8 @@ container-smoke: $(CONTAINER_ID)
 	docker logs $(shell cat $(CONTAINER_ID))
 
+prune: clean
+	docker system prune -f
```
Nit: This will remove all dangling containers/images/volumes/etc. on the user machine which may be an unexpected side-effect. I would try to limit the "clean" targets so they only remove stuff specific to the project.

For example, if I cloned this repo and ran "make prune" on my machine, it would remove all stopped containers which I might have intended to start again which would be unexpected. At the least it should give a warning or something I think.

wadells replied:
A good point.  I'll remove this target, as it isn't important to the project.

wadells replied:
Addressed by 067830ffffc201fb700a220f142541d8221c78fc

r0mant commented on Makefile:
```diff
@@ -1,22 +1,42 @@
 IMAGE_ID=$(DOCKERDIR)/.image_id
 CONTAINER_ID=$(DOCKERDIR)/.container_id
+IMAGE_TAR=$(PYDIR)/wellknown.tar
+DOCKER_SCHMUTZ=$(IMAGE_TAR) $(IMAGE_ID) $(CONTAINER_ID)
```
What's "schmutz"?

wadells replied:
It comes from Yiddish I believe, and generally means "mess" or "dirt".  E.g. you might say "Clean that schmutz off your face!" to a child sticky with ice cream.

I'm by no means attached to this particular phrasing if you feel something else is more understandable.

wadells committed:
Remove prune target.

Roman (r0mant) is rightfully concerned this target may clobber stuff
outside the project.  Since this target isn't necessary, we'll just
remove it and let folks figure out how to clean up stray containers on
their own.

Addresses https://github.com/javins/gravitational-audition/pull/4#discussion_r335759524
